File: modules/LoadedLookupTableTest.lua
dofile 'Asserts.lua'
dofile 'LoadedLookupTable.lua'

local lookupTable = nn.LoadedLookupTable.load()

-- Test the index and word reach the same location in the table
local embedding_from_index = lookupTable:forward(lookupTable:queryIndex('to'))[1]
local embedding_from_word = lookupTable:forward('to')[1]

assertEquals(embedding_from_index:size(1), embedding_from_word:size(1))
for i = 1, embedding_from_index:size(1) do
  assertEquals(embedding_from_index[i], embedding_from_word[i])
end

-- Test known indices
assertEquals(1739, nn.LoadedLookupTable.PADDING)
assertEquals(1740, nn.LoadedLookupTable.UNKNOWN)
assertEquals(nn.LoadedLookupTable.PADDING, lookupTable:queryIndex('PADDING'))
assertEquals(nn.LoadedLookupTable.UNKNOWN, lookupTable:queryIndex('UNKNOWN'))

assertEquals(1246, lookupTable:queryIndex('0/0/0'))
assertEquals(1643, lookupTable:queryIndex('0th'))
assertEquals(96969, lookupTable:queryIndex('reply'))
assertEquals(nn.LoadedLookupTable.UNKNOWN, lookupTable:queryIndex('wree'))
assertEquals(129976, lookupTable:queryIndex('zygmunt'))

-- Test known embeddings
local embedding1 = lookupTable:forward(1)[1]
assertFloatEquals(-1.03682, embedding1[1])
assertFloatEquals(1.5799, embedding1[5])
assertFloatEquals(-0.64853, embedding1[50])

local embedding130000 = lookupTable:forward(130000)[1]
assertFloatEquals(-0.238824, embedding130000[1])
assertFloatEquals(-1.76695, embedding130000[5])
assertFloatEquals(-1.4733, embedding130000[50])

-- Test back propagation
local zeroLookupTable = nn.LoadedLookupTable(
    torch.Tensor(5, 5):zero(), {['word'] = 1}, {[1] = 'word'})

local zeroEmbedding = zeroLookupTable:forward('word')[1]
assertFloatEquals(0, zeroEmbedding[1])
assertFloatEquals(0, zeroEmbedding[2])
assertFloatEquals(0, zeroEmbedding[3])
assertFloatEquals(0, zeroEmbedding[4])
assertFloatEquals(0, zeroEmbedding[5])

-- back propagate
zeroLookupTable:backwardUpdate('word', torch.Tensor(5, 5):fill(1), 0.1)

local zeroEmbedding = zeroLookupTable:forward('word')[1]
assertFloatEquals(-0.1, zeroEmbedding[1])
assertFloatEquals(-0.1, zeroEmbedding[2])
assertFloatEquals(-0.1, zeroEmbedding[3])
assertFloatEquals(-0.1, zeroEmbedding[4])
assertFloatEquals(-0.1, zeroEmbedding[5])

zeroLookupTable:reset(torch.Tensor(5, 5):zero())

-- reset weights
local zeroEmbedding = zeroLookupTable:forward('word')[1]
assertFloatEquals(0, zeroEmbedding[1])
assertFloatEquals(0, zeroEmbedding[2])
assertFloatEquals(0, zeroEmbedding[3])
assertFloatEquals(0, zeroEmbedding[4])
assertFloatEquals(0, zeroEmbedding[5])

-- back propagate
zeroLookupTable:backward('word', torch.Tensor(5, 5):fill(1))
zeroLookupTable:updateParameters(0.1)

local zeroEmbedding = zeroLookupTable:forward('word')[1]
assertFloatEquals(-0.1, zeroEmbedding[1])
assertFloatEquals(-0.1, zeroEmbedding[2])
assertFloatEquals(-0.1, zeroEmbedding[3])
assertFloatEquals(-0.1, zeroEmbedding[4])
assertFloatEquals(-0.1, zeroEmbedding[5])

File: theano_port/LoadedLookupTable_test.py
from LoadedLookupTable import LoadedLookupTable
import numpy
import unittest

class LoadedLookupTableTest(unittest.TestCase):
  
  @classmethod
  def setUpClass(cls):
    cls.lookup_table = LoadedLookupTable.load()

  def test_known_indices(self):
    self.assertEqual(1738, LoadedLookupTable.PADDING)
    self.assertEqual(1739, LoadedLookupTable.UNKNOWN)
    self.assertEqual(LoadedLookupTable.PADDING, self.lookup_table.query_index('PADDING'))
    self.assertEqual(LoadedLookupTable.UNKNOWN, self.lookup_table.query_index('UNKNOWN'))

    self.assertEqual('PADDING', self.lookup_table.query_word(LoadedLookupTable.PADDING))
    self.assertEqual('UNKNOWN', self.lookup_table.query_word(LoadedLookupTable.UNKNOWN))
    self.assertEqual('UNKNOWN', self.lookup_table.query_word(-1))

    self.assertEqual(1245, self.lookup_table.query_index('0/0/0'))
    self.assertEqual(1642, self.lookup_table.query_index('0th'))
    self.assertEqual(96968, self.lookup_table.query_index('reply'))
    self.assertEqual(LoadedLookupTable.UNKNOWN, self.lookup_table.query_index('wree'))
    self.assertEqual(129975, self.lookup_table.query_index('zygmunt'))

  def test_forward(self):
    embedding_from_index = self.lookup_table.forward(self.lookup_table.query_index('to'))
    embedding_from_word = self.lookup_table.forward('to')
    self.assertTrue(numpy.allclose(embedding_from_index, embedding_from_word))

    embedding0 = self.lookup_table.forward(0)
    self.assertAlmostEqual(-1.03682, embedding0[0])
    self.assertAlmostEqual(1.5799, embedding0[4])
    self.assertAlmostEqual(-0.64853, embedding0[49])

    embedding129999 = self.lookup_table.forward(129999)
    self.assertAlmostEqual(-0.238824, embedding129999[0])
    self.assertAlmostEqual(-1.76695, embedding129999[4])
    self.assertAlmostEqual(-1.4733, embedding129999[49])

  def test_backward(self):
    zero_lookup_table = LoadedLookupTable(numpy.zeros((5, 5)), {'word': 1}, {1: 'word'})
    zero_embedding = zero_lookup_table.forward('word')
    self.assertTrue(numpy.allclose(numpy.zeros((5, 5)), zero_embedding))
    zero_lookup_table.backward('word', numpy.ones((5,)))
    zero_lookup_table.update(0.1)
    expected = numpy.ones((5,))
    expected.fill(-0.1)
    zero_embedding = zero_lookup_table.forward('word')
    self.assertTrue(numpy.allclose(expected, zero_embedding))

  def test_backward_update(self):
    zero_lookup_table = LoadedLookupTable(numpy.zeros((5, 5)), {'word': 1}, {1: 'word'})
    zero_embedding = zero_lookup_table.forward('word')
    self.assertTrue(numpy.allclose(numpy.zeros((5, 5)), zero_embedding))
    zero_lookup_table.backward_update('word', numpy.ones((5,)), 0.1)
    expected = numpy.ones((5,))
    expected.fill(-0.1)
    zero_embedding = zero_lookup_table.forward('word')
    self.assertTrue(numpy.allclose(expected, zero_embedding))

if '__main__' == __name__:
  unittest.main()
